docs/guides/2sl2b-design.md

```diff
@@ -0,0 +1,96 @@
+
+The API framework Dagger uses should enable:
+
+1. Calling of automation from anywhere and any language
+   - The API may be invoked from a CLI command, from a script written in any of our supported languages, from an HTTP request sent by a browser, etc.
+1. Authoring of automation in any language
+   - The caller of an API doesn't need to know what language the API is implemented with.
+   - This freedom leaves both caller and callee to choose whichever language fits their particular job the best without having to think about interoperability.
+   - Rather than reimplementing the same automation in every possible programming language, it only needs to be implemented once and s then re-usable by all.
+   - Strengths of different languages thus enhance each other rather than creating silos.
+     - If Python has the world's greatest library for accomplishing X, an extension author should feel free to import+use it; they will be making that functionality available for anyone else in Go, Typescript, Bash and any other Dagger-supported language along the way.
+   - Additionally, two cooperating teams no longer necessarily have to come to consensus on a common language.
+     - One team can implement automation in Go and then provide it to another team that uses Javascript.
+1. Type-safety
+   - The inputs and output of APIs should all have well-defined types+shapes. Languages that have type checking should be able to use this to enable compile-time checks for devs calling and authoring extensions.
+1. Low-boilerplate
+   - The acts of A) defining an API and B) calling an API should not involve lots of boilerplate or domain-specific knowledge.
+
+Dagger uses GraphQL as its low-level language-agnostic API framework for achieving the above goals. The reasons for this are discussed more in the FAQ below.
+
+### Language-native feeling
+
+The Dagger SDK for each language should feel natural and intuitive to those already familiar with that language. It should embrace the strengths of the language and not fight any of its conventions. You should not need to become an expert in the Dagger SDK to use the Dagger SDK.
+
+There are many aspects to the above, but a few selected implications are:
+
+1. Defining an API should not feel very different to just defining a function in the language you choose to use. The conversion of those functions to language-agnostic API calls should be fully hidden behind the Dagger SDK.
+   - We call this feature "code-first schemas"
+1. Calling an API should not feel very different than just calling functions in the language you choose to use. The conversion of those function calls to language-agnostic API calls should be fully hidden behind the Dagger SDK.
+   - We support this through code-generated SDKs, which implement typed interfaces to extension APIs.
+
+## FAQ
```

Comment:
I do not know if that's accurate but that may help to add a section to explain how you can (will) connect Dagger 2.0 to Dagger 3.0.
If I understand well, Cue can just be another supported language, I think it's interesting to talk about that.